geospatial/src/Dem.cc

```diff
@@ -275,35 +272,71 @@ float Dem::MaxElevation() const
 bool Dem::GeoReference(double _x, double _y,
     gz::math::Angle &_latitude, gz::math::Angle &_longitude) const
 {
-  if (this->dataPtr->isNonEarthDem)
+  if (this->dataPtr->isUnknownDem)
   {
-    gzerr << "Can not retrieve WGS84 coordinates from non-Earth DEM."
+    gzerr << "Can not retrieve coordinates from unknown DEM."
             << std::endl;
     return false;
   }
 
   double geoTransf[6];
   if (this->dataPtr->dataSet->GetGeoTransform(geoTransf) == CE_None)
   {
+    OGRCoordinateTransformation *cT = nullptr;
+    double xGeoDeg, yGeoDeg;
     OGRSpatialReference sourceCs;
     OGRSpatialReference targetCs;
-    OGRCoordinateTransformation *cT;
-    double xGeoDeg, yGeoDeg;
 
-    // Transform the terrain's coordinate system to WGS84
-    const char *importString
-        = strdup(this->dataPtr->dataSet->GetProjectionRef());
-    if (importString == nullptr || importString[0] == '\0')
+    if (this->dataPtr->sphericalCoordinates.Surface() ==
+        math::SphericalCoordinates::EARTH_WGS84)
     {
-      gzdbg << "Projection coordinate system undefined." << std::endl;
-      return false;
+      // Transform the terrain's coordinate system to WGS84
+      const char *importString
+          = strdup(this->dataPtr->dataSet->GetProjectionRef());
+      if (importString == nullptr || importString[0] == '\0')
+      {
+        // LCOV_EXCL_START
+        gzdbg << "Projection coordinate system undefined." << std::endl;
+        return false;
+        // LCOV_EXCL_STOP
+      }
+      sourceCs.importFromWkt(&importString);
+      targetCs.SetWellKnownGeogCS("WGS84");
     }
-    sourceCs.importFromWkt(&importString);
-    targetCs.SetWellKnownGeogCS("WGS84");
+    else if (this->dataPtr->sphericalCoordinates.Surface() ==
+        math::SphericalCoordinates::MOON_SCS)
+    {
+      sourceCs = *(this->dataPtr->dataSet->GetSpatialRef());
+      targetCs = OGRSpatialReference();
+
+      std::string moonLatLongProjStr =
+        "+proj=latlong +a=1738100 +b=1736000";
+
+      targetCs.importFromProj4(moonLatLongProjStr.c_str());
+    }
+    else if (this->dataPtr->sphericalCoordinates.Surface() ==
+        math::SphericalCoordinates::CUSTOM_SURFACE)
+    {
+      sourceCs = *(this->dataPtr->dataSet->GetSpatialRef());
+      targetCs = OGRSpatialReference();
+
+      double axisEquatorial =
+        this->dataPtr->sphericalCoordinates.SurfaceAxisEquatorial();
+      double axisPolar =
+        this->dataPtr->sphericalCoordinates.SurfaceAxisPolar();
+
+      std::string customSurfaceLatLongProjStr =
+        "+proj=latlong +a=" + std::to_string(axisEquatorial) +
+        " +b=" + std::to_string(axisPolar);
+
+      targetCs.importFromProj4(customSurfaceLatLongProjStr.c_str());
+    }
```

Comment:
One more suggestion:
```suggestion
    else if (this->dataPtr->sphericalCoordinates.Surface() ==
        math::SphericalCoordinates::MOON_SCS ||
        this->dataPtr->sphericalCoordinates.Surface() ==
        math::SphericalCoordinates::CUSTOM_SURFACE)
    {
      sourceCs = *(this->dataPtr->dataSet->GetSpatialRef());
      targetCs = OGRSpatialReference();

      double axisEquatorial =
        this->dataPtr->sphericalCoordinates.SurfaceAxisEquatorial();
      double axisPolar =
        this->dataPtr->sphericalCoordinates.SurfaceAxisPolar();

      std::string customSurfaceLatLongProjStr =
        "+proj=latlong +a=" + std::to_string(axisEquatorial) +
        " +b=" + std::to_string(axisPolar);

      targetCs.importFromProj4(customSurfaceLatLongProjStr.c_str());
    }
```

Reply:
Done https://github.com/gazebosim/gz-common/pull/377/commits/ba01d65f784848e8e5735affb3f97b8be8a2760e